### PR TITLE
KAFKA-14585: Moving StorageTool from core to tools module

### PR DESCRIPTION
**Pull request by muralibasani**
*Description*
This PR does the below.
- Moves the classes StorageTool.scala and StorageToolTest.scala to tools module with Java conversions.
- There is a dependency by QuorumTestHarness & KafkaClusterTestKit on method 'formatCommand' in StorageTool. So this method has been moved to ToolUtils to avoid circular dependency on the modules. Pls suggest any alternative if possible.
- build.gradle has been updated to mention 2 more dependencies of tools module
- checkstyle import-control is updated to mention allowed imports by tools module

*Summary of testing strategy (including rationale)
Existing test StorageToolTest.scala to moved to tools modules as StorageToolTest.java

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by muralibasani** [comment]
@mimaison PR is ready to review. Possible for you to take a look ?
- Had to modify build.gradle to mention couple of dependencies. And to avoid circular dependency (as formatCommand is being used in other classes in core module.) had to duplicate that method into ToolsUtils. Can there be an alternative ?

- There is an unrelated test failing in a different module. Hope it's ok ?
[2023-03-21T12:41:00.523Z] FAILURE: Build failed with an exception.
[2023-03-21T12:41:00.523Z] 
[2023-03-21T12:41:00.523Z] * What went wrong:
[2023-03-21T12:41:00.523Z] Execution failed for task ':connect:api:integrationTest'.

**Comment by mimaison** [comment]
Thanks for the PR! We really want to avoid having tools depend on core. If there are methods from core that we want to use, we can move them to server-common.

**Comment by muralibasani** [comment]
> Thanks for the PR! We really want to avoid having tools depend on core. If there are methods from core that we want to use, we can move them to server-common.

If tools module should not depend on core module, then the below imports should be moved to the server-common. Ex : KafkaConfig constructor is being used, so need to move the whole class in that case

import kafka.server.BrokerMetadataCheckpoint;
import kafka.server.KafkaConfig;
import kafka.server.MetaProperties;
import kafka.server.RawMetaProperties;
import kafka.tools.TerseFailure;

this is a much bigger change then right ?

**Comment by mimaison** [comment]
Yes moving some of these tools is not trivial.
In this case:
- `kafka.tools.TerseFailure` can be replaced with `org.apache.kafka.tools.TerseException`
- For `KafkaConfig` I don't think it makes sense to move it but we could maybe extract some constants. We have a few `*Config` classes already (like `RaftConfig`, `LogConfig`, etc).
- I'm not too familiar with the other files but it looks like we should be able to move them

**Comment by muralibasani** [comment]
> * RaftConfig

@mimaison while continuing on this, moving constants to RaftConfig, creating classes like MetaProperties, RawMetaProperties looks okish.

But this one BrokerMetadataCheckpoint, not sure how to proceed. Can you take a deeper look into the dependency classes, just wondering if we are in the right track ?

**Comment by muralibasani** [comment]
@mimaison any suggestions ?

File: checkstyle/import-control.xml [hunk 1]
```diff
@@ -414,6 +414,11 @@
     <allow pkg="org.apache.kafka.clients.admin" />
     <allow pkg="org.apache.kafka.clients.producer" />
     <allow pkg="org.apache.kafka.clients.consumer" />
+    <allow pkg="org.apache.kafka.metadata.bootstrap" />
+    <allow pkg="org.apache.kafka.server.common" />
+    <allow pkg="kafka.server" />
+    <allow pkg="kafka.tools" />
```
Comment (fvaleri) [inline checkstyle/import-control.xml]: I see you already discussed this. We do not want tools to have a dependency on core.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        formatParser.addArgument("--cluster-id", "-t").action(store()).required(true).help("The cluster ID to use.");
+        formatParser.addArgument("--ignore-formatted", "-g").action(storeTrue());
+        formatParser.addArgument("--release-version", "-r").action(store()).help(String.format("A KRaft release version to use for the initial metadata version. The minimum is 3.0, the default is %s", MetadataVersion.latest().version()));
+
+        return parser.parseArgsOrFail(args);
+    }
+
+    static List<String> configToLogDirectories(KafkaConfig kafkaConfig) {
+        Seq<String> logDirs = kafkaConfig.logDirs();
+        SortedSet<String> directories = new TreeSet<>(JavaConverters.asJavaCollection(logDirs));
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: `JavaConverters` is deprecated, you should use `CollectionConverters` instead.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                stream.println("");
+                return 1;
+            } else {
+                return 0;
+            }
+        }
+    }
+
+
+    static Namespace parseArguments(String... args) throws Exception {
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: Exception is never thrown in this method.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        } catch (Exception exception) {
+            System.err.println(exception.getMessage());
+            Exit.exit(1);
+        }
+    }
+
+    private static void executeCommand(Namespace namespace, String command, Optional<KafkaConfig> config) throws Exception {
+        switch (command) {
+            case "info": {
+                if (config.isPresent()) {
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: This check is repeated two times. Why not checking it at the start of the method when command is not `random-uuid`. Even better would be to do this in a parsing and validation phase. In other tools, we create a private static inner class for that. You could also handle `configToLogDirectories` there, and avoid the double parsing.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                    List<String> directories = configToLogDirectories(config.get());
+                    String clusterId = namespace.getString("cluster_id");
+                    MetadataVersion metadataVersion = getMetadataVersion(namespace, Optional.of(config.get().interBrokerProtocolVersionString()));
+                    if (!metadataVersion.isKRaftSupported()) {
+                        throw new TerseFailure("Must specify a valid KRaft metadata version of at least 3.0.");
+                    }
+                    MetaProperties metaProperties = buildMetadataProperties(clusterId, config.get());
+                    Boolean ignoreFormatted = namespace.getBoolean("ignore_formatted");
+                    if (!configToSelfManagedMode(config.get())) {
+                        throw new TerseFailure("The kafka configuration file appears to be for " + "a legacy cluster. Formatting is only supported for clusters in KRaft mode.");
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: Do we miss a CR here?

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+    static int formatCommand(PrintStream stream, List<String> directories, MetaProperties metaProperties, MetadataVersion metadataVersion, boolean ignoreFormatted) {
+        if (directories.isEmpty()) {
+            throw new TerseFailure("No log directories found in the configuration.");
+        }
+        List<String> unformattedDirectories = directories.stream().filter(directory -> {
+            if (!Files.isDirectory(Paths.get(directory)) || !Files.exists(Paths.get(directory, "meta.properties"))) {
+                return true;
+            } else if (!ignoreFormatted) {
+                throw new TerseFailure("Log directory " + directory + " is already formatted. " + "Use --ignore-formatted to ignore this directory and format the others.");
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: Another missing CR, I guess.

**Review by fvaleri — CHANGES_REQUESTED** [review]
Hi @muralibasani, thanks for working on this.

You still need to update `bin/kafka-storage.sh` and `bin/windows/kafka-storage.bat`.

About the `BrokerMetadataCheckpoint`, all classes there seem to be well isolated, and I think they could be moved to the metadata module (for example, in a new package named `org.apache.kafka.metadata.broker`). @ijuma can you advice here?

**Comment by muralibasani** [comment]
@fvaleri  @ijuma 
Managed to remove the core dependency from tools. Added org.apache.kafka.metadata.broker
If am in the right direction, will add tests, fix conflicts, and several other minor things.

**Comment by fvaleri** [comment]
Thanks @muralibasani, let me know when you are ready for another review.

**Comment by muralibasani** [comment]
> Thanks @muralibasani, let me know when you are ready for another review.

@fvaleri made the necessary changes I believe. Pls take a look. Thanks.
- last test in StorageToolsTest was not working as expected due to io write error (BrokerMetadataCheckpoint line 61 atomicMoveWithFallback fails.   [here](https://github.com/muralibasani/kafka/blob/KAFKA-14585/metadata/src/main/java/org/apache/kafka/metadata/broker/BrokerMetadataCheckpoint.java#L61) ), hence changed the exit status to 1, instead of 0 [here](https://github.com/muralibasani/kafka/blob/KAFKA-14585/tools/src/test/java/org/apache/kafka/tools/StorageToolTest.java#L382).
- Copied certain methods of StorageTools.scala to ToolsUtils.scala so there is no circular dependency
- LogConfig changes to be verified with extra care

**Review by fvaleri — COMMENTED** [review]
@muralibasani thanks, this is just a first pass. 

Looks like there are still some checkstyle errors to fix and a couple of errors on KafkaClusterTestKit. Can you also avoid moving imports around and other formatting changes as it makes the review harder. Should we use a Map instead of introducing Pair?

**Comment by muralibasani** [comment]
> @muralibasani thanks, this is just a first pass.
> 
> Looks like there are still some checkstyle errors to fix and a couple of errors on KafkaClusterTestKit. Can you also avoid moving imports around and other formatting changes as it makes the review harder. Should we use a Map instead of introducing Pair?

@fvaleri fixed checkstyle errors, imports and KafkaClusterTestKit. Regarding Pair, I can change to Map, but in scala it more reflects a Pair. Need to fix my IDE for checkstyle, formatting keeps changing if I open the file.

File: checkstyle/import-control.xml [hunk 7]
```diff
@@ -302,6 +302,11 @@
     <allow pkg="org.apache.kafka.clients.admin" />
     <allow pkg="org.apache.kafka.clients.producer" />
     <allow pkg="org.apache.kafka.clients.consumer" />
+    <allow pkg="org.apache.kafka.metadata.bootstrap" />
+    <allow pkg="org.apache.kafka.metadata.broker" />
+    <allow pkg="org.apache.kafka.storage.internals.log" />
+    <allow pkg="kafka.tools" />
+    <allow pkg="kafka.utils" />
```
Comment (fvaleri) [inline checkstyle/import-control.xml]: This entry is duplicated, and we don't need it if you get rid of kafka.utils.TestUtils dependency in StorageToolTest (it's only two simple methods).

Then, we can get rid of both kafka.utils and kafka.tools.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+            stream.println("No directories specified.");
+            return 0;
+        } else {
+            if (!foundDirectories.isEmpty()) {
+                if (foundDirectories.size() == 1) {
+                    stream.println("Found log directory:");
+                } else {
+                    stream.println("Found log directories:");
+                }
+                foundDirectories.forEach(d -> stream.printf(d + "%n"));
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: In the old implementation we have 2 spaces before the log dir path. It is important to keep the same output, because there may be some tool parsing it that we don't want to break.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+                stream.println("");
+            }
+
+            if (!problems.isEmpty()) {
+                if (problems.size() == 1) {
+                    stream.println("Found problem:");
+                } else {
+                    stream.println("Found problems:");
+                }
+                problems.forEach(d -> stream.printf(d + "%n"));
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java]: In the old implementation we have 2 spaces before the problem desc. It is important to keep the same output, because there may be some tool parsing it that we don't want to break.

**Review by fvaleri — COMMENTED** [review]
Hi, the behavior is different when I pass an invalid config file:

```sh
### OLD
$ bin/kafka-storage.sh info -c ~/.local/tmp/kafka/server2/config/log4j.properties
Exception in thread "main" org.apache.kafka.common.config.ConfigException: Missing required configuration `zookeeper.connect` which has no default value.
	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2183)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:2174)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1580)
	at kafka.tools.StorageTool$.$anonfun$main$1(StorageTool.scala:50)
	at scala.Option.flatMap(Option.scala:283)
	at kafka.tools.StorageTool$.main(StorageTool.scala:50)
	at kafka.tools.StorageTool.main(StorageTool.scala)

$ bin/kafka-storage.sh format -c ~/Downloads/test/wrong.properties -t "$(bin/kafka-storage.sh random-uuid)"
The kafka configuration file appears to be for a legacy cluster. Formatting is only supported for clusters in KRaft mode.

### NEW
$ bin/kafka-storage.sh info -c ~/.local/tmp/kafka/server2/config/log4j.properties
Found problem:
/tmp/kafka-logs does not exist

$ bin/kafka-storage.sh format -c ~/Downloads/test/wrong.properties -t "$(bin/kafka-storage.sh random-uuid)"
The node.id must be set to a non-negative integer. We saw -1
```

I still need to look the rest of the code. Will update later.

File: checkstyle/import-control.xml (hunk 7 above)
Comment (muralibasani) [reply checkstyle/import-control.xml]: Nice one. removed.

**Comment by muralibasani** [comment]
> Hi, the behavior is different when I pass an invalid config file:
> 
> ```shell
> ### OLD
> $ bin/kafka-storage.sh info -c ~/.local/tmp/kafka/server2/config/log4j.properties
> Exception in thread "main" org.apache.kafka.common.config.ConfigException: Missing required configuration `zookeeper.connect` which has no default value.
> 	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2183)
> 	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:2174)
> 	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1580)
> 	at kafka.tools.StorageTool$.$anonfun$main$1(StorageTool.scala:50)
> 	at scala.Option.flatMap(Option.scala:283)
> 	at kafka.tools.StorageTool$.main(StorageTool.scala:50)
> 	at kafka.tools.StorageTool.main(StorageTool.scala)
> 
> $ bin/kafka-storage.sh format -c ~/Downloads/test/wrong.properties -t "$(bin/kafka-storage.sh random-uuid)"
> The kafka configuration file appears to be for a legacy cluster. Formatting is only supported for clusters in KRaft mode.
> 
> ### NEW
> $ bin/kafka-storage.sh info -c ~/.local/tmp/kafka/server2/config/log4j.properties
> Found problem:
> /tmp/kafka-logs does not exist
> 
> $ bin/kafka-storage.sh format -c ~/Downloads/test/wrong.properties -t "$(bin/kafka-storage.sh random-uuid)"
> The node.id must be set to a non-negative integer. We saw -1
> ```
> 
> I still need to look the rest of the code. Will update later.

@fvaleri I see it's more complicated. Having this kind of validation method change https://github.com/apache/kafka/blob/cc4e699d4cb2880d05603e2e8310d28e1c9f201a/core/src/main/scala/kafka/server/KafkaConfig.scala#L2181 in LogConfig introduces several other test failures of different classes like [MetadataQuorumCommandTest and FeatureCommandTest. What do you suggest ?

**Comment by muralibasani** [comment]
@fvaleri code fixed based on review comments.
- new LogConfig constructor added
- new method to validate zk and broker config
- new test with zk config fail

**Comment by muralibasani** [comment]
> Hi, the behavior is different when I pass an invalid config file:
> 
> ```shell
> ### OLD
> $ bin/kafka-storage.sh info -c ~/.local/tmp/kafka/server2/config/log4j.properties
> Exception in thread "main" org.apache.kafka.common.config.ConfigException: Missing required configuration `zookeeper.connect` which has no default value.
> 	at kafka.server.KafkaConfig.validateValues(KafkaConfig.scala:2183)
> 	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:2174)
> 	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:1580)
> 	at kafka.tools.StorageTool$.$anonfun$main$1(StorageTool.scala:50)
> 	at scala.Option.flatMap(Option.scala:283)
> 	at kafka.tools.StorageTool$.main(StorageTool.scala:50)
> 	at kafka.tools.StorageTool.main(StorageTool.scala)
> 
> $ bin/kafka-storage.sh format -c ~/Downloads/test/wrong.properties -t "$(bin/kafka-storage.sh random-uuid)"
> The kafka configuration file appears to be for a legacy cluster. Formatting is only supported for clusters in KRaft mode.
> 
> ### NEW
> $ bin/kafka-storage.sh info -c ~/.local/tmp/kafka/server2/config/log4j.properties
> Found problem:
> /tmp/kafka-logs does not exist
> 
> $ bin/kafka-storage.sh format -c ~/Downloads/test/wrong.properties -t "$(bin/kafka-storage.sh random-uuid)"
> The node.id must be set to a non-negative integer. We saw -1
> ```
> 
> I still need to look the rest of the code. Will update later.

these tests should be ok now. fixed. had to make changes similar to KafkaConfig

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+            Optional<LogConfig> config = Optional.ofNullable(namespace.getString("config")).map(p -> {
+                try {
+                    return new LogConfig(Utils.loadProps(p), Collections.emptySet(), true);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+            executeCommand(namespace, command, config);
+        } catch (Exception exception) {
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:78]: Given we use TerseException, I think it's better to have the following error handling.

```sh
        } catch (TerseException e) {
            System.err.println(e.getMessage());
            Exit.exit(1);
        } catch (Throwable e) {
            System.err.println(e.getMessage());
            System.err.println(Utils.stackTrace(e));
            Exit.exit(1);
        }
```

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java [hunk 11]
```diff
@@ -389,10 +462,23 @@ public LogConfig(Map<?, ?> props, Set<String> overriddenConfigs) {
...
+        this.nodeId = getInt(LogConfig.NODE_ID_PROP);
+        this.brokerIdGenerationEnable = getBoolean(BROKER_ID_GENERATION_ENABLE_PROP);
+        this.logDirs = getString(LOG_DIRS_PROP);
+        this.logDir = getString(LOG_DIR_PROP);
+        processRoles = parseProcessRoles();
+        this.zkConnect = getString(ZOOKEEPER_CONNECT_PROP);
         remoteLogConfig = new RemoteLogConfig(this, retentionMs, retentionSize);
     }
 
+    public LogConfig(Map<?, ?> props, Set<String> overriddenConfigs, boolean populateSynonymsAndValidateProps) {
```
Comment (fvaleri) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:477]: populateSynonymsAndValidateProps is never used

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+            System.err.println(exception.getMessage());
+            Exit.exit(1, exception.getMessage());
+        }
+    }
+
+    private static void executeCommand(Namespace namespace, String command, Optional<LogConfig> config) throws Exception {
+        final String info = "info";
+        final String format = "format";
+        if ((command.equals(info) || command.equals(format)) && !config.isPresent()) {
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:87]: This is useless, because argparse4j validates all required args when parsing, which is before this point. This also means that we don't need the two variables above.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+            directories.add(metadataLogDir);
+        }
+        return new ArrayList<>(directories);
+    }
+
+    static boolean configToSelfManagedMode(LogConfig logConfig) {
+        return !logConfig.parseProcessRoles().isEmpty();
+    }
+
+
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:267]: Nit: extra empty line.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+                problems.forEach(d -> stream.println("  " + d));
+                stream.println("");
+                return 1;
+            } else {
+                return 0;
+            }
+        }
+    }
+
+
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:231]: Nit: extra empty line.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 15]
```diff
@@ -0,0 +1,479 @@
...
+            default:
+                throw new RuntimeException("Unknown command " + command);
+        }
+    }
+
+    static int infoCommand(PrintStream stream, boolean selfManagedMode, List<String> directories) throws IOException {
+        List<String> problems = new ArrayList<>();
+        List<String> foundDirectories = new ArrayList<>();
+        Optional<RawMetaProperties> prevMetadata = Optional.empty();
+        for (String directory : directories) {
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:142]: Thanks for removing the double sorting contained in `directories.sorted.foreach`.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+                        }
+                    }
+                }
+            }
+        }
+
+        if (prevMetadata.isPresent()) {
+            if (selfManagedMode) {
+                if (prevMetadata.get().getVersion() == 0) {
+                    problems.add("The kafka configuration file appears to be for a cluster in KRaft mode, but " + "the directories are formatted for legacy mode.");
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:185]: Nit: remove string concatenation or add a carriage return.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 17]
```diff
@@ -0,0 +1,479 @@
...
+                }
+            } else if (prevMetadata.get().getVersion() == 1) {
+                problems.add("The kafka configuration file appears to be for a legacy cluster, but " + "the directories are formatted for a cluster in KRaft mode.");
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:188]: Nit: remove string concatenation or add a carriage return.

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 18]
```diff
@@ -0,0 +1,479 @@
...
+        }
+
+        switch (command) {
+            case info: {
+                List<String> directories = configToLogDirectories(config.get());
+                boolean selfManagedMode = configToSelfManagedMode(config.get());
+                Exit.exit(infoCommand(System.out, selfManagedMode, directories));
+                break;
+            }
+            case format: {
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:98]: This code looks good, but I was wondering that we could improve it by moving all checks and object creations to the command methods. Something like:

```sh
        switch (command) {
            case info: {
                Exit.exit(infoCommand(System.out, config));
                break;
            }
            case format: {
                Exit.exit(formatCommand(System.out, namespace, config));
                break;
            }
...
```

File: tools/src/main/java/org/apache/kafka/tools/StorageTool.java [hunk 19]
```diff
@@ -0,0 +1,479 @@
...
+
+    public static BootstrapMetadata buildBootstrapMetadata(MetadataVersion metadataVersion,
+                                                           Optional<List<ApiMessageAndVersion>> metadataOptionalArguments,
+                                                           String source) {
+        List<ApiMessageAndVersion> metadataRecords = new ArrayList<>();
+        metadataRecords.add(new ApiMessageAndVersion(new FeatureLevelRecord()
+            .setName(MetadataVersion.FEATURE_NAME)
+            .setFeatureLevel(metadataVersion.featureLevel()), (short) 0));
+
+        metadataOptionalArguments.ifPresent(metadataArguments -> metadataArguments.forEach(metadataRecords::add));
```
Comment (fvaleri) [inline tools/src/main/java/org/apache/kafka/tools/StorageTool.java:359]: Use can use addAll here.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java [hunk 20]
```diff
@@ -489,6 +575,29 @@ public static void validateValues(Map<?, ?> props) {
         }
     }
 
+    public static Map<?, ?> populateAndValidateProps(Map<?, ?> props) {
```
Comment (fvaleri) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:578]: I think we can find a better name or merge with validateOtherConfig. Wdyt?

File: streams/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java [hunk 21]
```diff
@@ -38,6 +37,7 @@
 import java.util.Map;
 import java.util.Properties;
 import java.util.Set;
+import java.util.stream.Stream;
```
Comment (mimaison) [inline streams/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java:40]: This is not related to the rest of this PR, let's undo this change

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java [hunk 22]
```diff
@@ -59,6 +35,35 @@
...
+import static org.apache.kafka.common.config.ConfigDef.Range.atLeast;
+import static org.apache.kafka.common.config.ConfigDef.Range.between;
+import static org.apache.kafka.common.config.ConfigDef.Type.BOOLEAN;
+import static org.apache.kafka.common.config.ConfigDef.Type.DOUBLE;
+import static org.apache.kafka.common.config.ConfigDef.Type.INT;
+import static org.apache.kafka.common.config.ConfigDef.Type.LIST;
+import static org.apache.kafka.common.config.ConfigDef.Type.LONG;
+import static org.apache.kafka.common.config.ConfigDef.Type.STRING;
+import static org.apache.kafka.common.config.ConfigDef.ValidString.in;
+import static org.apache.kafka.server.common.MetadataVersion.IBP_3_0_IV1;
```
Comment (mimaison) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:65]: Can we undo reordering the imports?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java [hunk 23]
```diff
@@ -230,6 +235,55 @@ public Optional<String> serverConfigName(String configName) {
...
+    private static final String NODE_ID_DOC = "The node ID associated with the roles this process is playing when `process.roles` is non-empty. " +
+        "This is required configuration when running in KRaft mode.";
+    private static final String PROCESS_ROLES_DOC = "The roles that this process plays: 'broker', 'controller', or 'broker,controller' if it is both. " +
+        "This configuration is only applicable for clusters in KRaft (Kafka Raft) mode (instead of ZooKeeper). Leave this config undefined or empty for ZooKeeper clusters.";
+
+    private static final String BROKER_ID_GENERATION_ENABLE_DOC = "Enable automatic broker id generation on the server. When enabled the value configured for $MaxReservedBrokerIdProp should be reviewed.";
+
+    private static final String MAX_RESERVED_BROKER_ID_DOC = "Max number that can be used for a broker.id";
+
+    private static final String ZK_CONNECT_DOC = "Specifies the ZooKeeper connection string in the form <code>hostname:port</code> where host and port are the " +
```
Comment (fvaleri) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:282]: I don't think this information should be in LogConfig.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java [hunk 24]
```diff
@@ -230,6 +235,55 @@ public Optional<String> serverConfigName(String configName) {
         FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG
     ));
 
+    private static final String DEFAULT_LOG_DIR = "/tmp/kafka-logs";
```
Comment (mimaison) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:238]: This class is for configurations for the storage layer. We should not put unrelated configurations like `ZOOKEEPER_CONNECT_PROP` in it.

File: checkstyle/import-control.xml [hunk 25]
```diff
@@ -302,6 +302,9 @@
     <allow pkg="org.apache.kafka.clients.admin" />
     <allow pkg="org.apache.kafka.clients.producer" />
     <allow pkg="org.apache.kafka.clients.consumer" />
+    <allow pkg="org.apache.kafka.metadata.bootstrap" />
+    <allow pkg="org.apache.kafka.metadata.broker" />
+    <allow pkg="org.apache.kafka.storage.internals.log" />
```
Comment (mimaison) [inline checkstyle/import-control.xml:307]: We should not import internals packages.

File: core/src/main/scala/kafka/utils/ToolsUtils.scala [hunk 26]
```diff
@@ -79,4 +94,226 @@ object ToolsUtils {
...
+      setFeatureLevel(metadataVersion.featureLevel()), 0.toShort));
+
+    metadataOptionalArguments.foreach { metadataArguments =>
+      for (record <- metadataArguments) metadataRecords.add(record)
+    }
+
+    BootstrapMetadata.fromRecords(metadataRecords, source)
+  }
+
+  def parseArguments(args: Array[String]): Namespace = {
```
Comment (mimaison) [inline core/src/main/scala/kafka/utils/ToolsUtils.scala:114]: It looks like these methods are specific to the Kafka storage tool. We should not put them in a utils class and not in core.

File: metadata/src/main/java/org/apache/kafka/metadata/broker/RawMetaProperties.java [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+
+package org.apache.kafka.metadata.broker;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+
+public class RawMetaProperties {
```
Comment (mimaison) [inline metadata/src/main/java/org/apache/kafka/metadata/broker/RawMetaProperties.java:26]: I wonder if we should first do a PR to move these classes from core to metadata.

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. I left a few comments.

**Review by fvaleri — CHANGES_REQUESTED** [review]
Hi @muralibasani, thanks for the additional work. I left some comments.

As @mimaison, I also think that it would be better to create a dedicated PR for metadata and storage classes, to be merged before this one. I also think that methods migrated to LogConfig should replace the ones in KafkaConfig, so that we can test them properly.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java (hunk 24 above)
Comment (muralibasani) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:238]: @mimaison scala version of StorageTool executes this code block https://github.com/apache/kafka/blob/4981fa939d588645401619bfc3e321dc523d10e7/core/src/main/scala/kafka/server/KafkaConfig.scala#L2181C5-L2181C25 Hence I added similar in LogConfig. wdyt ?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java (hunk 23 above)
Comment (muralibasani) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:282]: @fvaleri scala version of StorageTool executes this code block https://github.com/apache/kafka/blob/4981fa939d588645401619bfc3e321dc523d10e7/core/src/main/scala/kafka/server/KafkaConfig.scala#L2181C5-L2181C25 Hence I added similar in LogConfig. wdyt ? So property has to be defined first and then we can perform validations.

**Comment by muralibasani** [comment]
> Hi @muralibasani, thanks for the additional work. I left some comments.
> 
> As @mimaison, I also think that it would be better to create a dedicated PR for metadata and storage classes, to be merged before this one. I also think that methods migrated to LogConfig should replace the ones in KafkaConfig, so that we can test them properly.

@fvaleri @mimaison to be sure,  I shall create another PR keeping KafkaConfig usage in StorageTool. And this PR for replacing KafkaConfig with LogConfig changes. Right ?

**Comment by fvaleri** [comment]
@muralibasani the new PR should contain the work you already did with LogConfig, RawMetaProperties, etc. Note that we can't keep two versions of the same code without the risk of divergence, so every class/method in every module where it is used should call the Java version in the end. Once the new PR is merged, we can rebase this one and complete the tool migration. Make sense?

**Review by fvaleri — COMMENTED** [review]
@muralibasani the new PR should contain the work you already did with LogConfig, RawMetaProperties, etc. Note that we can't keep two versions of the same code without the risk of divergence, so every class/method in every module where it is used should call the Java version in the end. Once the new PR is merged, we can rebase this one and complete the tool migration. Make sense?

**Comment by muralibasani** [comment]
> _Description_
> This PR does the below.

@fvaleri here is the new PR https://github.com/apache/kafka/pull/14106 There are a few dependencies on the scala version of BrokerMetadaCheckpoint, but we can update them after StorageTools are migrated in a different PR?

**Comment by muralibasani** [comment]
As discussed with @fvaleri , closing this PR. Thank you.
